odbc/src/api/data.rs

```diff
@@ -194,26 +497,209 @@ pub unsafe fn format_and_return_bson(
+            SqlReturn::ERROR
         }
-        CDataType::Time | CDataType::TypeTime => {
-            let dt = data.to_date();
+    }
+}
+
+pub unsafe fn format_time(
+    mongo_handle: &mut MongoHandle,
+    col_num: USmallInt,
+    target_value_ptr: Pointer,
+    str_len_or_ind_ptr: *mut Len,
+    data: Bson,
+) -> SqlReturn {
+    let stmt = (*mongo_handle).as_statement().unwrap();
+    let mut guard = stmt.write().unwrap();
+    let dt = data.to_date();
+    match dt {
+        Ok(dt) => {
             let data = Time {
                 hour: dt.hour() as u16,
                 minute: dt.minute() as u16,
                 second: dt.second() as u16,
             };
-            set_output_fixed_data(&data, target_value_ptr, str_len_or_ind_ptr)
+            set_output_fixed_data(
+                &data,
+                col_num,
+                guard.var_data_indicies.as_mut().unwrap(),
+                target_value_ptr,
+                str_len_or_ind_ptr,
+            )
         }
-        CDataType::Date | CDataType::TypeDate => {
-            let dt = data.to_date();
+        Err(e) => {
+            guard.errors.push(e);
+            SqlReturn::ERROR
+        }
+    }
+}
+
+pub unsafe fn format_date(
+    mongo_handle: &mut MongoHandle,
+    col_num: USmallInt,
+    target_value_ptr: Pointer,
+    str_len_or_ind_ptr: *mut Len,
+    data: Bson,
+) -> SqlReturn {
+    let stmt = (*mongo_handle).as_statement().unwrap();
+    let mut guard = stmt.write().unwrap();
+    let dt = data.to_date();
+    match dt {
+        Ok(dt) => {
             let data = Date {
                 year: dt.year() as i16,
                 month: dt.month() as u16,
                 day: dt.day() as u16,
             };
-            set_output_fixed_data(&data, target_value_ptr, str_len_or_ind_ptr)
+            set_output_fixed_data(
+                &data,
+                col_num,
+                guard.var_data_indicies.as_mut().unwrap(),
+                target_value_ptr,
+                str_len_or_ind_ptr,
+            )
+        }
+        Err(e) => {
+            guard.errors.push(e);
+            SqlReturn::ERROR
+        }
+    }
+}
+
+pub unsafe fn format_and_return_bson(
+    mongo_handle: &mut MongoHandle,
+    col_num: USmallInt,
+    target_type: CDataType,
+    target_value_ptr: Pointer,
+    buffer_len: Len,
+    str_len_or_ind_ptr: *mut Len,
+    data: Bson,
+) -> SqlReturn {
+    // If the data is null or undefined we immediately return NULL_DATA indicator.
+    match data {
+        Bson::Null | Bson::Undefined => {
+            if str_len_or_ind_ptr.is_null() {
+                mongo_handle.add_diag_info(ODBCError::IndicatorVariableRequiredButNotSupplied);
+                return SqlReturn::SUCCESS_WITH_INFO;
+            }
+            let stmt = (*mongo_handle).as_statement().unwrap();
+            let mut guard = stmt.write().unwrap();
+            let indicies = guard.var_data_indicies.as_mut().unwrap();
```

Comment:
One question, where does the indices get reset to `0`?  As in for when the next row is fetched and the data in the column has not yet been retrieved.  
It may be something to do when calling to fetch the next row.